proposed/http-handlers/request-handlers-meta.md

```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+this system:
+
+- Producing a fallback response if no other layer does.
+- Dispatching middleware.
+
+#### Reusable Middleware Examples
+
+In the examples above, we have two middleware composed in each. In order for
+these to work in either situation, we need to write them such that they interact
+appropriately.
+
+Generally speaking, implementors of middleware should follow these guidelines:
+
+- Test the request for a required condition. If it does not satisfy that
+  condition, use a composed prototype response or a composed response factory
+  to generate and return a response.
+
+- If pre-conditions are met, delegate creation of the response to the provided
+  request handler, optionally providing a "new" request by manipulating the
+  provided request (e.g., `$handler->handle($request->withAttribute('foo',
+  'bar')`).
+
+- Either pass the response returned by the request handler unaltered, or provide
+  a new response by manipulating the one returned (e.g., `return
+  $response->withHeader('X-Foo-Bar', 'baz')`).
+
+The `CorsMiddleware` is one that will exercise all three of these guidelines: if
+certain request conditions are not met, it needs to return an error response. In
+the case of a "pre-flight" request, it will return a response immediately with
+details for the client. In a case where the pre-flight request has already been
+made and expected headers are present, it delegates to the request handler, and
+then injects headers into the returned response.
+
+```php
+class CorsMiddleware implements MiddlewareInterface
+{
+    private $analyzer;
+    private $responsePrototype;
+
+    public function __construct(CorsAnalyzer $analyzer, ResponseInterface $responsePrototype)
+    {
+        $this->analyzer = $analyzer;
+        $this->responsePrototype = $responsePrototype;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) : ResponseInterface
+    {
+        $corsResult = $this->analyzer->analyze($request);
+
+        if ($corsResult->isInvalid()) {
+            return $corsResult->prepareForbiddenResponse($this->responsePrototype);
+        }
+
+        if ($corsResult->isPreFlight()) {
+            return $corsResult->preparePreFlightResponse($this->responsePrototype);
+        }
+
+        $response = $handler->handle($request);
+
+        return $corsResult->isCorsScoped()
+            ? $corsResult->prepareResponse($this->responsePrototype)
+            : $response;
+    }
+}
+```
+
+Note the use of a "response prototype" in the above example; this approach
+allows the middleware to be agnostic of which PSR-7 implementation is in use,
+and yet still return a response in cases where it determines it should not
+delegate to the handler.Similarly, it is not concerned with how the request
+handler is implemented; it merely uses it to produce a response when
+pre-conditions have been met.
+
+The `RoutingMiddleware` implementation described below follows a similar
```

Comment:
Yes this is better, but I think the process should be:

- If routing matches, continue processing.
- Else return a `HTTP Not Found` response.

Reply:
That's actually exactly how we do it in Expressive. :+1:

The only issue I see with that is that it complicates the example: I then need to add dispatch middleware and a response prototype or factory. I felt the "map route to handler" approach was sufficiently simple and didn't require extra explanation.